pressurecooker/images.py

```diff
@@ -21,50 +25,117 @@
 from pdf2image import convert_from_path
 from PIL import Image, ImageOps
 
-def create_tiled_image(source_images, fpath_out):
+from le_utils.constants import file_formats
+
+from .thumbscropping import scale_and_crop
+
+
+
+# SMARTCROP UTILS
+################################################################################
+
+THUMBNAIL_SIZE = (400, 225)  # 16:9
+
+def scale_and_crop_thumbnail(image, size=THUMBNAIL_SIZE, crop="smart", **kwargs):
     """
-    Create a tiled image from list of image paths provided in source_images and
-    write result to fpath_out.
+    Scale and crop the PIL Image ``image`` to maximum dimensions of ``size``.
+    By default, ``crop`` is set to "smart" which will crop the image down to size
+    based on the entropy content of the pixels. The other options are:
+    * Use ``crop="0,0"`` to crop from the left and top edges
+    * Use ``crop=",0"`` to crop from the top edge.
+    Optional keyword arguments:
+    * ``zoom=X``: crop outer X% before starting
+    * ``target``: recenter here before cropping (default center ``(50, 50)``)
+    See the ``scale_and_crop`` docs in ``thumbscropping.py`` for more details.
     """
-    root = math.sqrt(len(source_images))
-    assert len(source_images) > 0, "Must provide at least one image to tile"
-    assert int(root + 0.5) ** 2 == len(source_images), "Number of images must be a perfect square"
+    return scale_and_crop(image, size, crop=crop, upscale=True, **kwargs)
+
 
-    images = list(map(Image.open, source_images))
-    widths, heights = zip(*(i.size for i in images))
-
-    max_dimension = int(min(max(widths), max(heights)))
-    offset = int(max_dimension / int(root))
-
-    new_im = Image.new('RGB', (max_dimension, max_dimension))
-
-    index = x_index = y_index = 0
-    root = int(root)
-    while y_index < root and index < len(images):
-        x_index = 0
-        while x_index < root and index < len(images):
-            im = ImageOps.fit(images[index], (int(offset), int(offset)), Image.ANTIALIAS)
-            new_im.paste(im, (int(offset * x_index), int(offset * y_index)))
-            x_index += 1
-            index += 1
-        y_index += 1
-    new_im.save(fpath_out)
 
-def create_image_from_pdf_page(fpath_in, fpath_out, page_number=0):
+# THUMBNAILS FOR CONTENT KINDS
+################################################################################
+
+def create_image_from_epub(epubfile, fpath_out, crop=None):
+    """
+    Generate a thumbnail image from `epubfile` and save it to `fpath_out`.
+    Raises ThumbnailGenerationError if thumbnail extraction fails.
+    """
+    book = ebooklib.epub.read_epub(epubfile)
+    # 1. try to get cover image from book metadata (content.opf)
+    cover_item = None
+    covers = book.get_metadata('http://www.idpf.org/2007/opf', 'cover')
+    if covers:
+        cover_tuple = covers[0] # ~= (None, {'name':'cover', 'content':'item1'})
+        assert cover_tuple[1]['name'] == 'cover', 'wrong key name'
```

Comment:
What does it mean if this happens? Is this a sign of an invalid epub file, or could there be multiple cover tags, but with one of them not being the cover?